arangod/Cluster/ActionBase.cpp

```diff
@@ -0,0 +1,261 @@
+  _actionStarted = secs_since_epoch();
+  return false;
+}
+
+void ActionBase::complete() {
+  _actionDone = secs_since_epoch();
+  auto cf = ApplicationServer::getFeature<ClusterFeature>("Cluster");
+  if (cf != nullptr) {
+    cf->syncDBServerStatusQuo();
+  }
+  _state = COMPLETE;
+}
+
+void ActionBase::fail() {
+  _actionDone = secs_since_epoch();
+  auto cf = ApplicationServer::getFeature<ClusterFeature>("Cluster");
+  if (cf != nullptr) {
+    cf->syncDBServerStatusQuo();
+  }
+  _state = FAILED;
+}
+
+/// @brief execution finished successfully or failed ... and race timer expired
+bool ActionBase::done() const {
+
+  return (COMPLETE==_state || FAILED==_state) &&
+    _actionDone + _feature.getSecondsActionsBlock() <= secs_since_epoch();
+
+} // ActionBase::done
+
+ActionDescription const& ActionBase::describe() const {
+  return _description;
+}
+
+MaintenanceFeature& ActionBase::feature() const {
+  return _feature;
+}
+
+VPackSlice const ActionBase::properties() const {
+  return _description.properties()->slice();
+}
+
+
+/// @brief Initiate a new action that will start immediately, pausing this action
+void ActionBase::createPreAction(std::shared_ptr<ActionDescription> const & description) {
+
+  _preAction = description;
+  _feature.preAction(description);
+
+  // shift from EXECUTING to WAITINGPRE ... EXECUTING is set to block other
+  //  workers from picking it up
+  if (_preAction) {
+    setState(WAITINGPRE);
+  } else {
+    _result.reset(TRI_ERROR_BAD_PARAMETER, "preAction rejected parameters.");
+  } // else
+
+} // ActionBase::createPreAction
+
+
+/// @brief Retrieve pointer to action that should run before this one
+std::shared_ptr<Action> ActionBase::getPreAction() {
+  return (_preAction != nullptr) ? _feature.findAction(_preAction) : nullptr;
+}
+
+
+/// @brief Retrieve pointer to action that should run after this one
+std::shared_ptr<Action> ActionBase::getPostAction() {
+  return (_postAction != nullptr) ? _feature.findAction(_postAction) : nullptr;
+}
+
+
+/// @brief Create a new action that will start after this action successfully completes
+void ActionBase::createPostAction(std::shared_ptr<ActionDescription> const& description) {
+
+  // preAction() sets up what we need
+  _postAction = description;
+  _feature.postAction(description);
+
+  // shift from EXECUTING to WAITINGPOST ... EXECUTING is set to block other
+  //  workers from picking it up
+  if (_postAction) {
+    setState(WAITINGPOST);
+  } else {
+    _result.reset(TRI_ERROR_BAD_PARAMETER, "preAction rejected parameters for _postAction.");
+  } // else
+
+} // ActionBase::createPostAction
+
+
+void ActionBase::startStats() {
+
+// done when first  _actionStarted = std::chrono::system_clock::now();
+
+} // ActionBase::startStats
+
+
+void ActionBase::incStats() {
+
+  ++_progress;
+  //_actionLastStat = std::chrono::system_clock::now();
+
+} // ActionBase::incStats
+
+
+void ActionBase::endStats() {
+
+  //_actionDone = std::chrono::system_clock::now();
```

Comment:
This needs to be reactivated.  MaintenanceWorker::nextState() uses endStats().  nextState() properly marks all Actions (including postActions) with _state and _actionDone based upon _result code.  This is to ensure coder of Actions is only required to manage result object.

Reply:
fixed